addons/mrp/models/product.py

```diff
@@ -70,10 +82,26 @@ class ProductProduct(models.Model):
         compute='_compute_used_in_bom_count', compute_sudo=False)
     mrp_product_qty = fields.Float('Manufactured',
         compute='_compute_mrp_product_qty', compute_sudo=False)
+    is_kits = fields.Boolean(compute="_compute_bom_count", compute_sudo=False)
 
     def _compute_bom_count(self):
+        domain = ['|', ('product_id', 'in', self.ids), '&', ('product_id', '=', False), ('product_tmpl_id', 'in', self.product_tmpl_id.ids)]
+        group_data = self.env['mrp.bom'].read_group(domain=domain, fields=['product_id', 'product_tmpl_id', 'type'], groupby=['product_id', 'product_tmpl_id', 'type'], lazy=False)
+        template_mapping = defaultdict(lambda: [0, False]) # (#_of_bom, _has_phantom)
+        product_mapping = defaultdict(lambda: [0, False]) # (#_of_bom, _has_phantom)
+        for group in group_data:
+            product_id = group['product_tmpl_id'][0]
+            product_tmpl_id = group['product_tmpl_id'][0]
+            bom_type = group['type'][0]
+            if product_id:
+                product_mapping[product_id][0] += 1
+                product_mapping[product_id][1] = bom_type == "phantom"
+            else:
+                template_mapping[product_tmpl_id][0] += 1
+                template_mapping[product_tmpl_id][1] = bom_type == "phantom"
         for product in self:
-            product.bom_count = self.env['mrp.bom'].search_count(['|', ('product_id', '=', product.id), '&', ('product_id', '=', False), ('product_tmpl_id', '=', product.product_tmpl_id.id)])
+            product.bom_count = product_mapping[product.id][0] + template_mapping[product.product_tmpl_id.id][0]
```

Comment:
`product.bom_count = product_mapping.get(product.id, [])[0] + template_mapping.get(product.product_tmpl_id.id, [])[0]`

use get() for product without bom. if not it will raise.